src/viewer/ViewerView.js

```diff
@@ -1,146 +1,53 @@
-        }
-      });
-    };
-    this.scene.add(InputHandler);
-
-    // input init
-    // default to right hand.
-    // avoid XRInputs data structures due to XRPK oninputsourcechange bug
-    this.primaryController = Renderer.xr.getController(0);
-    this.scene.add(this.primaryController);
-    this.CreateSkybox();
-    this.CreateSelectionOrb();
-  }
+    State.eventHandler.addEventListener(
+      "setinitposition",
+      this.setInitPosition.bind(this)
+    );
 
-  HandlePhotoSelection(e, button) {
-    // xrpk way
-    XRInput.inputSources.forEach((inputSource, i) => {
-      if (e.handedness === inputSource.handedness) {
-        this.primaryIndex = i;
-      }
-    });
-    this.primaryController = Renderer.xr.getController(this.primaryIndex);
-
-    let handInSphere = this.SphereDistanceTest(this.primaryController.position, this.orbRadius)
-    if (handInSphere)
-      this.CyclePhoto();
-  }
-
-  SphereDistanceTest(pos, dist) {
-    if (!this.selectionOrb) return;
-
-    let controllerPos = new Vector3(pos.x, pos.y, pos.z);
-    let spherePos = new Vector3();
-    this.selectionOrb.getWorldPosition(spherePos);
-    let d = controllerPos.sub(spherePos);
-
-    return (d.x * d.x + d.y * d.y + d.z * d.z) < Math.pow(dist, 2);
-  }
-
-  LoadPhoto(data) {
-    this.Select(data.photoIndex);
-  }
-
-  CreateSkybox() {
-    let geometry = new SphereBufferGeometry(500, 180, 180);
-    geometry.scale(- 1, 1, 1);
-
-    let photoTexture = this.photos[this.currentPhoto];
-    this.skyboxMaterial = new MeshBasicMaterial({
-        map: photoTexture,
-        precision: "highp",
-    });
-    this.skybox = new Mesh(geometry, this.skyboxMaterial);
-    this.scene.add(this.skybox);
-  };
-
-  CreateSelectionOrb() {
-    let photoTexture = this.photos[this.NextPhoto()];
-
-    let geometry = new SphereGeometry(this.orbRadius, 20, 20);
-    geometry.scale(- 1, 1, 1);
-    let material = new MeshBasicMaterial({
-      color: 0xffffff,
-      map: photoTexture,
-    });
-    let sphere = new Mesh( geometry, material );
-    sphere.position.setY(1.2);
-    sphere.position.setZ(-1);
+    this.scene = new Scene();
+    loadScene(this.scene);
 
-    this.selectionOrb = sphere;
-    this.scene.add(this.selectionOrb);
+    this.ui = new UI(this.scene);
+    const input = new Input(this.scene, this.ui);
   }
-
-  Select(index) {
-    this.currentPhoto = index;
-
-    this.skybox.material.map = this.photos[this.currentPhoto];
-    this.selectionOrb.material.map = this.photos[this.NextPhoto()]
+  incrementPhotoIndex(increment) {
+    this.currentPhotoIndex = this.ui.retrieveNextPhotoIndex(increment);
+    const data = { photoIndex: this.currentPhotoIndex };
+    this.publish("model", Q.INCREMENT_PHOTO_INDEX, data);
   }
-
-  CyclePhoto() {
-    this.Select(this.NextPhoto());
-    let data = { photoIndex: this.currentPhoto };
-    this.publish("viewer", "remoteselectphoto", data);
+  selectPhoto(data) {
+    State.eventHandler.dispatchEvent("selectphoto", data);
```

Comment:
Why not just call `this.ui.selectPhoto(data)` directly since the class already has a reference to UI? You could call that function directly from the subscribe on line 21

Reply:
nice catch! originally component comms were all event based without passing references - this makes much more sense for the current structure.